### PR TITLE
Oleg Gondar Homework Module 5 - Annotations and Generics corrected wi…

### PR DESCRIPTION
**Pull request by GondarOleg**
…th Sonar

### CODE REVIEW

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/Run.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    public static Object createClass(String className) throws ReflectiveOperationException {
+
+        Class<?> c = null;
+        try {
+            c = Class.forName(className);
+            Constructor<?> constructor = c.getConstructor();
+            return constructor.newInstance();
+        } catch (ClassNotFoundException | NoSuchMethodException | InstantiationException | IllegalAccessException e) {
+            System.out.println(e.getMessage());
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/Run.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+                } catch (ReflectiveOperationException e) {
+
+                    org.junit.Assert.assertTrue(e.getCause() instanceof MyException);
+                    System.out.println("Exception throwing test passed!");
+
+                }
+
+            } else if (method.isAnnotationPresent(Test.class) && method.getAnnotation(Test.class).ignore()) {
+                System.out.println("Method: " + method.getName() + " not tested");
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/Run.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    private Run() {
+    }
+
+    public static void main(String[] args) throws ReflectiveOperationException {
+
+
+        try {
+            performActorTest("edu.task1.mytestingclasses.Actor");
+        } catch ( ReflectiveOperationException  e) {
+            System.out.println(e.getMessage());
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/Run.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+        for (Method method :
+                actor.getClass().getMethods()) {
+            if (method.isAnnotationPresent(Test.class) && !method.getAnnotation(Test.class).ignore()) {
+
+                try {
+
+                    method.invoke(actor);
+
+                } catch (ReflectiveOperationException e) {
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/Run.java]: CRITICAL Either log or rethrow this exception. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS1166)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/Run.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+                    org.junit.Assert.assertTrue(e.getCause() instanceof MyException);
+                    System.out.println("Exception throwing test passed!");
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/mytestingclasses/Actor.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    public void testException() throws MyException {
+
+        throw new MyException("!!!!!!!!!!!!!!!!");
+
+    }
+
+    @Test(ignore = true)
+    public void ignoredTest() {
+
+        System.out.println("Ignored method");
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/mytestingclasses/Actor.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/mytestingclasses/Actor.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        System.out.println("Actor first name test passed");
+
+    }
+
+    @Test()
+    public void testSetActorLastName() {
+        Actor a = new Actor();
+        a.setActorLastName("Last");
+        Assert.assertTrue("Last".equals(a.getActorLastName()));
+        System.out.println("Actor last name test passed");
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/mytestingclasses/Actor.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/mytestingclasses/Actor.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        this.actorLastName = actorLastName;
+    }
+
+
+    @Test()
+    public void testSetActorFirstName() {
+        Actor a = new Actor();
+        a.setActorFirstName("First");
+        Assert.assertTrue("First".equals(a.getActorFirstName()));
+        System.out.println("Actor first name test passed");
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/mytestingclasses/Actor.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task3/Run.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    private Run() {
+    }
+
+    public static void main(String[] args) {
+
+        Unit<Integer> unit = new Unit<>(1);
+        System.out.println(unit);
+
+        Pair<Integer, String> pair = new Pair<>(1, "two");
+        System.out.println(pair);
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task3/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task3/Run.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+        Triplet<Integer, String, Integer> triplet = new Triplet<>(1, "second", 3);
+        System.out.println(triplet);
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task3/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task3/Run.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+ */
+public class Run {
+
+    private Run() {
+    }
+
+    public static void main(String[] args) {
+
+        Unit<Integer> unit = new Unit<>(1);
+        System.out.println(unit);
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task3/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    public static <T extends Comparable> void runMethods(T[] t) {
+
+        System.out.println("Finding values for: " + t.getClass().getSimpleName());
+        System.out.println("Max is: " + CompareUtils.minMax(t, false).getValue());
+        System.out.println("Min is: " + CompareUtils.minMax(t, true).getValue());
+        System.out.println("Median is: " + CompareUtils.median(t).getValue());
+        System.out.println("____________________________________________________");
+        System.out.println("Work with comparator");
+        System.out.println("Finding values for: " + t.getClass().getSimpleName());
+        System.out.println("Max is: " + CompareUtils.sortWithComparator(t, new MyComparator(), "max").getValue());
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+        System.out.println("Min is: " + CompareUtils.sortWithComparator(t, new MyComparator(), "min").getValue());
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+        System.out.println("Median is: " + CompareUtils.sortWithComparator(t, new MyComparator(), "median").getValue());
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+        System.out.println("____________________________________________________");
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+
+        runMethods(s);
+        runMethods(i);
+
+
+    }
+
+    public static <T extends Comparable> void runMethods(T[] t) {
+
+        System.out.println("Finding values for: " + t.getClass().getSimpleName());
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+        System.out.println("Max is: " + CompareUtils.minMax(t, false).getValue());
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+        System.out.println("Min is: " + CompareUtils.minMax(t, true).getValue());
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+        System.out.println("Median is: " + CompareUtils.median(t).getValue());
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+        System.out.println("____________________________________________________");
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+        System.out.println("Work with comparator");
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+        System.out.println("Finding values for: " + t.getClass().getSimpleName());
```
Comment (technicaltaskjava) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task2/Run.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://localhost:9000/coding_rules#rule_key=squid%3AS106)

**Comment by technicaltaskjava** [comment]
SonarQube analysis reported 22 issues:
* CRITICAL 1 critical
* MAJOR 21 major

Watch the comments in this conversation to review them.

File: Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/Run.java (hunk 4 above)
Comment (GondarOleg) [inline Oleg_Gondar/Module 5 - Annotations and Generics/src/main/java/edu/task1/Run.java]: I will log all exceptions in my next homework. I must test assertion in this task (test excpect in annotation).

**Pull request closed without merging**